src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
 
     // SFC methods
     @Override
-    public NetworkElement registerNetworkElement(List<NetworkElement> inspectedPorts) throws Exception {
-        return null;
+    public NetworkElement registerNetworkElement(List<NetworkElement> inspectionPorts) throws Exception {
+
+        ServiceFunctionChainEntity sfc = new ServiceFunctionChainEntity();
+        //check for null or empty list
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "Port Pair Group memeber list");
+
+        this.utils.addPortPairGroupToServiceFunction(inspectionPorts, sfc);
+        return this.txControl.required(() -> {
+            return this.em.merge(sfc);
+        });
+
     }
 
     @Override
-    public NetworkElement updateNetworkElement(NetworkElement portGroup, List<NetworkElement> inspectedPorts)
+    public NetworkElement updateNetworkElement(NetworkElement portGroup, List<NetworkElement> inspectionPorts)
             throws Exception {
-        // no-op
-        return null;
+
+        this.utils.throwExceptionIfNullElementAndId(portGroup, "Port Pair Group ServiceFunctionChain Id");
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "Port Pair Group update memeber list");
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(portGroup.getElementId());
+
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Update failed",
+                    portGroup.getElementId()));
+        }
+
+        sfc.getPortPairGroups().clear();
+        this.utils.addPortPairGroupToServiceFunction(inspectionPorts, sfc);
+
+        return this.txControl.required(() -> {
+            return this.em.merge(sfc);
+        });
     }
 
     @Override
-    public void deleteNetworkElement(NetworkElement portGroupId) throws Exception {
-        // no-op
+    public void deleteNetworkElement(NetworkElement serviceFunctionChainId) throws Exception {
+
+        this.utils.throwExceptionIfNullElementAndId(serviceFunctionChainId, "ServiceFunctionChain Id");
+
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(serviceFunctionChainId.getElementId());
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Delete failed",
+                    serviceFunctionChainId.getElementId()));
+        }
+
+        this.txControl.required(() -> {
+            //need to delete sfc if found in a transaction so fetch it again
+            ServiceFunctionChainEntity sfcToDel = this.em.find(ServiceFunctionChainEntity.class, serviceFunctionChainId.getElementId());
```

Comment:
move the findBySfcId method here

Reply:
The unit test is failing with exception if we are throwing a exception in transaction.
So I have to opt for a check before entering into the transaction. Please Let me know if there is another way.

Reply:
hmm...we are throwing in transaction in other places....what  is different here? lets discuss tommorow.

Reply:
used try, catch and it seems to be working. Please check and let me know.
we can throw an exception using a msg or I can call the same util API to throw the exception. Which were is preferred..

Reply:
i've tested locally. Exceptions thrown within transactions are wrapped with ScopedWorkException looks like...
So you should catch this exception and unwrap if required